cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +192,184 @@ constexpr bool is_hex_digit(char ch)
  */
-struct url_decode_esc_position_filter {
-  device_span<int32_t const> const d_offsets{};
-
-  /**
-   * @brief Detects if an escape sequence crosses a string boundary
-   *
-   * Returns true for an escape sequence that straddles a boundary between two strings.
-   *
-   * @param esc_pos_idx Character position corresponding to the start of an escape sequence
-   * @return true The escape sequence crosses a string boundary
-   * @return false The escape sequence does not cross a string boundary
-   */
-  __device__ bool operator()(size_type esc_pos_idx) const
-  {
-    // find the end offset of the current string
-    size_type const* offset_ptr =
-      thrust::upper_bound(thrust::seq, d_offsets.begin(), d_offsets.end(), esc_pos_idx);
-    return esc_pos_idx + 2 >= *offset_ptr;
+template <int threadblock_size, int block_size>
+__global__ void url_decode_char_counter(char const* const in_chars,
+                                        offset_type const* const in_offsets,
+                                        offset_type* const out_counts,
+                                        size_type const num_strings)
+{
+  constexpr int warps_per_threadblock = threadblock_size / cudf::detail::warp_size;
+  __shared__ char temporary_buffer[warps_per_threadblock][block_size + 2];
+  __shared__ typename cub::WarpReduce<int8_t>::TempStorage cub_storage[warps_per_threadblock];
+
+  int global_thread_id  = blockIdx.x * blockDim.x + threadIdx.x;
+  int global_warp_id    = global_thread_id / cudf::detail::warp_size;
+  int local_warp_id     = threadIdx.x / cudf::detail::warp_size;
+  int warp_lane         = threadIdx.x % cudf::detail::warp_size;
+  int nwarps            = gridDim.x * blockDim.x / cudf::detail::warp_size;
+  char* in_chars_shared = temporary_buffer[local_warp_id];
+
+  // Loop through strings, and assign each string to a warp.
+  for (size_type row_idx = global_warp_id; row_idx < num_strings; row_idx += nwarps) {
+    auto in_chars_string          = in_chars + in_offsets[row_idx];
+    auto string_length            = in_offsets[row_idx + 1] - in_offsets[row_idx];
+    int nblocks                   = (string_length + block_size - 1) / block_size;
+    offset_type escape_char_count = 0;
+
+    for (int block_idx = 0; block_idx < nblocks; block_idx++) {
+      int string_length_block = std::min(block_size, string_length - block_size * block_idx);
+
+      // Each warp collectively loads input characters of the current block to the shared memory.
+      // Two halo cells with 0s are added after the input characters to avoid branches when testing
+      // escape characters.
+      for (int char_idx = warp_lane; char_idx < string_length_block + 2;
+           char_idx += cudf::detail::warp_size) {
+        char ch    = 0;
+        int in_idx = block_idx * block_size + char_idx;
+        if (in_idx < string_length) ch = in_chars_string[in_idx];
+        in_chars_shared[char_idx] = ch;
```

Comment:
This is probably negligible (and honestly somewhat surprising to me that it's not optimized out), but a quick look at a minimal example in godbolt suggests that we get slightly more efficient PTX by hinting to the compiler to optimize out the `ch` variable
```suggestion
        int in_idx = block_idx * block_size + char_idx;
        in_chars_shared[char_idx] = (in_idx < string_length) ? in_chars_string[in_idx] : 0;
```

Reply:
Yes, I think using ternary operator is clearer, accepted this in https://github.com/rapidsai/cudf/pull/8622/commits/1e30938b5979055b582aaface178b0c357bd2018.

I play around with godbolt but I don't see any difference between the two:
https://godbolt.org/z/vxY1rTejv
https://godbolt.org/z/MPzMe7r9n